### DISCUSSION TITLE
Issue with utemplate - silent death of process

### DISCUSSION THREAD

**Discussion by JohannesMaierhofer** [miguelgrinberg/microdot#226] [Q&A]
Hi, I have some issues with the reliability of using the utemplate rendering. The thing is: 1 out of 10 times it works, and the next call could again silently crash the application. Observing the webbrowser, the rendered template always loads, but the referenced static files are often not loaded, or only loaded partially. Working with send_file, all the static files load perfectly well.

Pressing "stop" in Thonny unveils the following message:
```
GET /dataviewer 200

───────────────────────────────────────────────────
Traceback (most recent call last):
  File "<stdin>", line 123, in <module>
  File "microdot/microdot.py", line 1281, in run
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in wait_io_event
KeyboardInterrupt: 
```

Here are some code-snippets:
```
app = Microdot()
Response.default_content_type = 'text/html'

@app.route('/dataviewer', methods=['GET'])
async def dataviewer(request):
    filename = request.args['filename']
    if filename and filename.endswith('.json'):
        return Template('dataview.html').generate(filename=filename)
    else:
        return 'File not found', 404

@app.route('/static/<path:path>')
async def static(request, path):
    if '..' in path:
        # directory traversal is not allowed
        return 'Not found', 404
    return send_file('static/' + path)

if __name__ == '__main__':
    app.run(port=80, debug=True)
```

```
...     
<link type="text/css" rel="stylesheet" href="static/index.css">
      <script type="text/javascript" src="static/chart.umd.js"></script>
      <script type="text/javascript" src="static/WebAppDataView.js"></script>
...
```

Any help or comment is appreciated.

**Comment by miguelgrinberg** [comment]
The only thing that comes to mind is if you may be using templates that require more RAM than what you have in your device to compile. I'm honestly not familiar with the internals of utemplate, which isn't my project. Does the problem occur if you run your application on the Unix port of MicroPython, for example? It would be interesting to know.

**Reply by JohannesMaierhofer** [reply]
Thanks! I don't think it has to do with RAM (at least not for the utemplate), as the .html has only 25lines, one parameter. So not much to do. The rendered file from the template is always served. The subsequent static files are the ones making trouble. 
What I can't grasp: Why isn't it always the same behavior?

 I didn't setup a Unix port up to now, I will try that later...

**Reply by miguelgrinberg** [reply]
I'm sorry, then I do not understand the problem. If the template loads fine, then you need to look at the browser's console to see what error(s) the static files produce.

**Reply by JohannesMaierhofer** [reply]
sorry, it is the `send_file('static/' + path)` which crashes randomly after Template('dataview.html').generate(filename=filename) is called once.

**Reply by miguelgrinberg** [reply]
Crashes how? Please include the stack trace of the crash.

**Reply by JohannesMaierhofer** [reply]
I prepared a minimal working example, which shows my issue with minimal code and super small files. I tried with Micropython on ESP32S3 v 1.21, 1.22, 1.23. everywhere the same behavior. 
[firmwareMinimalExample.zip](https://github.com/miguelgrinberg/microdot/files/15150367/firmwareMinimalExample.zip)

**Reply by miguelgrinberg** [reply]
I don't know what I'm supposed to look for in this example. It runs fine for me. What error(s) are you getting?

**Reply by JohannesMaierhofer** [reply]
Sorry I bothered you. It seems I have a general issue with the filesystem and therefore the files are/were not served. It is not an issue of microdot!
